contracts/tatum/Tatum721Provenance.sol

```diff
@@ -0,0 +1,233 @@
+        returns (bool)
+    {
+        return super.supportsInterface(interfaceId);
+    }
+
+    function tokenURI(uint256 tokenId)
+        public
+        view
+        virtual
+        override(ERC721, ERC721URIStorage)
+        returns (string memory)
+    {
+        return ERC721URIStorage.tokenURI(tokenId);
+    }
+
+    function getTokenData(uint256 tokenId)
+        public
+        view
+        virtual
+        returns (string[] memory)
+    {
+        return _tokenData[tokenId];
+    }
+
+    function caller() public view virtual returns (address) {
+        return _msgSender();
+    }
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal virtual override(ERC721, ERC721Enumerable) {
+        super._beforeTokenTransfer(from, to, tokenId);
+    }
+
+    function _burn(uint256 tokenId)
+        internal
+        virtual
+        override(ERC721, ERC721URIStorage)
+    {
+        return ERC721URIStorage._burn(tokenId);
+    }
+
+    function tokenCashbackValues(uint256 tokenId)
+        public
+        view
+        virtual
+        returns (uint256[] memory)
+    {
+        return _cashbackValues[tokenId];
+    }
+
+    function tokenCashbackRecipients(uint256 tokenId)
+        public
+        view
+        virtual
+        returns (address[] memory)
+    {
+        return _cashbackRecipients[tokenId];
+    }
+
+    function updateCashbackForAuthor(uint256 tokenId, uint256 cashbackValue)
+        public
+        returns (bool)
+    {
+        for (uint256 i = 0; i < _cashbackValues[tokenId].length; i++) {
+            if (_cashbackRecipients[tokenId][i] == _msgSender()) {
+                _cashbackValues[tokenId][i] = cashbackValue;
+                return true;
+            }
+        }
+        return true;
+    }
+
+    function burn(uint256 tokenId) public virtual {
+        //solhint-disable-next-line max-line-length
+        require(
+            _isApprovedOrOwner(_msgSender(), tokenId),
+            "ERC721Burnable: caller is not owner nor approved"
+        );
+        _burn(tokenId);
+    }
+    function uintToString(uint v) private pure returns (string memory) {
+        uint maxlength = 100;
+        bytes memory reversed = new bytes(maxlength);
+        uint i = 0;
+        while (v != 0) {
+            uint remainder = v % 10;
+            v = v / 10;
+            reversed[i++] = bytes1(uint8(48 + remainder));
+        }
+        bytes memory s = new bytes(i); // i + 1 is inefficient
+        for (uint j = 0; j < i; j++) {
+            s[j] = reversed[i - j - 1]; // to avoid the off-by-one error
+        }
+        string memory str = string(s);  // memory isn't implicitly convertible to storage
+        return str;
+    }
+    function safeTransfer(
+        address to,
+        uint256 tokenId,
+        string memory data,
+        uint256 value
+    ) public payable {
+        string memory dataValue=uintToString(value);
```

Comment:
remove value from parameters, this line as well. you will pass to data the final value you are building here.